### PR TITLE
chore(ngc): refactor out tsc-wrapped

### PR DESCRIPTION
**Pull request by alexeagle**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-format
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)

**What is the new behavior?**

**Does this PR introduce a breaking change?**
- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

This allows angular's build to depend on some extensions, but not on code generation, and breaks a cycle in the angular build

### CODE REVIEW

File: tools/tsc-wrapped/src/main.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
+#!/usr/bin/env node
+
+// Must be imported first, because angular2 decorators throws on load.
+import 'reflect-metadata';
```
Comment (alexeagle) [inline tools/tsc-wrapped/src/main.ts]: this should not be needed in tsc-wrapped - let ngc do it

File: tools/tsc-watch/index.ts [hunk 2]
```diff
@@ -93,8 +93,7 @@ if (platform == 'node') {
     error: 'error',
     complete: 'Compilation complete. Watching for file changes.',
     onChangeCmds: [
-      ['node', 'dist/tools/cjs-jasmine/index-tools', '--', 'ts-metadata-collector/**/*{_,.}spec.js'],
-      ['node', 'dist/tools/cjs-jasmine/index-tools', '--', 'public_api_guard/**/*{_,.}spec.js']
+      ['node', 'dist/tools/cjs-jasmine/index-tools', '--', '{public_api_guard,tsc-wrapped}/**/*{_,.}spec.js']
```
Comment (tbosch) [inline tools/tsc-watch/index.ts]: If you keep `ts-metadata-collector`, you should keep the tests running...

File: tools/tsc-wrapped/test/tsc.spec.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import * as ts from 'typescript';
+import {Tsc} from '../src/tsc';
+
+describe('options parsing', () => {
```
Comment (tbosch) [inline tools/tsc-wrapped/test/tsc.spec.ts]: Would be nice to move some of the integration tests from `@angular/compiler_cli` to here that only tests tsickle and metadata production (e.g. https://github.com/angular/angular/blob/master/modules/%40angular/compiler_cli/integrationtest/test/basic_spec.ts#L11)

File: tools/tsc-watch/index.ts (hunk 2 above)
Comment (alexeagle) [inline tools/tsc-watch/index.ts]: removed in next commit

File: test.sh [hunk 4]
```diff
@@ -5,5 +5,6 @@ set -e -o pipefail
 cd `dirname $0`
 export NODE_PATH=$NODE_PATH:$(pwd)/dist/all:$(pwd)/dist/tools
 $(npm bin)/tsc -p tools
-node dist/tools/tsc-watch/ node watch
+# node dist/tools/tsc-watch/ node watch
+node dist/tools/tsc-watch/ tools watch
```
Comment (tbosch) [inline test.sh]: Maybe revert this change?

File: tools/ts-metadata-collector/package.json [hunk 5]
```diff
@@ -1,16 +0,0 @@
-{
```
Comment (tbosch) [inline tools/ts-metadata-collector/package.json]: We probably also should deprecate the `ts-metadata-collector` package on npm (or drop it)

File: test.sh (hunk 4 above)
Comment (alexeagle) [inline test.sh]: oops

File: tools/tsc-wrapped/test/tsc.spec.ts (hunk 3 above)
Comment (alexeagle) [inline tools/tsc-wrapped/test/tsc.spec.ts]: Yes, but the expense is having another integration test setup. I don't think we have a generalized way to run integration tests right now?

File: modules/@angular/compiler_cli/index.ts [hunk 6]
```diff
@@ -1,3 +1,2 @@
 export {CodeGenerator} from './src/codegen';
 export {NodeReflectorHost} from './src/reflector_host';
-export {TsickleHost, MetadataWriterHost} from './src/compiler_host';
```
Comment (tbosch) [inline modules/@angular/compiler_cli/index.ts]: I think we should still re export everyhing from `tsc-wrapped/index.ts` here. With that, our users just depend on `@angular/compiler_cli` and `tsc-wrapped` is just our private dependency.

File: tools/ts-metadata-collector/package.json (hunk 5 above)
Comment (alexeagle) [inline tools/ts-metadata-collector/package.json]: yes, I'll do the deprecation dance (already did it for `angular2-template-compiler` package)

File: tools/tsc-wrapped/README.md [hunk 7]
```diff
@@ -0,0 +1,12 @@
+# tsc-wrapped
+
+This is a wrapper around TypeScript's `tsc` program that allows us to hook in extra extensions.
```
Comment (tbosch) [inline tools/tsc-wrapped/README.md]: Maybe add: "This is an internal package used by `@angular/compiler_cli`. Please use that instead.

File: modules/@angular/compiler_cli/index.ts (hunk 6 above)
Comment (alexeagle) [inline modules/@angular/compiler_cli/index.ts]: agreed

File: tools/tsc-wrapped/README.md (hunk 7 above)
Comment (alexeagle) [inline tools/tsc-wrapped/README.md]: done

File: build.sh [hunk 8]
```diff
@@ -35,14 +35,8 @@ cd -
 TSCONFIG=./modules/tsconfig.json
 echo "====== (all)COMPILING: \$(npm bin)/tsc -p ${TSCONFIG} ====="
 # compile ts code
-# TODO: Right now we have a cycle in that the compiler_cli depends on Angular
```
Comment (tbosch) [inline build.sh]: You also need to fix https://github.com/angular/angular/blob/master/scripts/ci-lite/build.sh#L14

File: modules/tsconfig.json [hunk 9]
```diff
@@ -14,7 +14,8 @@
       "selenium-webdriver": ["./@angular/typings/selenium-webdriver/selenium-webdriver.d.ts"],
       "rxjs/*": ["../node_modules/rxjs/*"],
       "@angular/*": ["./@angular/*"],
-      "ts-metadata-collector": ["../dist/tools/ts-metadata-collector"]
+      "ts-metadata-collector": ["../dist/tools/ts-metadata-collector"],
```
Comment (chuckjaz) [inline modules/tsconfig.json]: Do we still need this line?

File: build.sh (hunk 8 above)
Comment (alexeagle) [inline build.sh]: done

**Comment by alexeagle** [comment]
merged in 4c26397
